Dockerfile

```diff
@@ -25,13 +26,52 @@ RUN GOARCH=$(echo $TARGETPLATFORM | cut -f2 -d '/') GCE_PD_CSI_STAGING_VERSION=$
 
 # MAD HACKS: Build a version first so we can take the scsi_id bin and put it somewhere else in our real build
 FROM k8s.gcr.io/build-image/debian-base:buster-v1.9.0 as mad-hack
-RUN clean-install udev
+RUN clean-install udev                                                                                               
 
 # Start from Kubernetes Debian base
-FROM k8s.gcr.io/build-image/debian-base:buster-v1.9.0
-COPY --from=builder /go/src/sigs.k8s.io/gcp-compute-persistent-disk-csi-driver/bin/gce-pd-csi-driver /gce-pd-csi-driver
+FROM k8s.gcr.io/build-image/debian-base:buster-v1.9.0 as debian
 # Install necessary dependencies
-RUN clean-install util-linux e2fsprogs mount ca-certificates udev xfsprogs
-COPY --from=mad-hack /lib/udev/scsi_id /lib/udev_containerized/scsi_id
+RUN clean-install util-linux e2fsprogs mount ca-certificates udev xfsprogs 
 
-ENTRYPOINT ["/gce-pd-csi-driver"]
+# Since we're leveraging apt to pull in dependencies, we use `gcr.io/distroless/base` because it includes glibc
+FROM gcr.io/distroless/base-debian11
+# Copy necessary dependencies into distoless base.
+COPY --from=builder /go/src/sigs.k8s.io/gcp-compute-persistent-disk-csi-driver/bin/gce-pd-csi-driver /gce-pd-csi-driver
+COPY --from=mad-hack /lib/udev/scsi_id /lib/udev_containerized/scsi_id
+COPY --from=debian /bin/umount /bin/umount
+COPY --from=debian /bin/mount /bin/mount
+COPY --from=debian /sbin/resize2fs /sbin/resize2fs
+COPY --from=debian /sbin/e2mmpstatus /sbin/e2mmpstatus
+COPY --from=debian /sbin/e2mmpstatus /sbin/e2mmpstatus
+COPY --from=debian /sbin/e* /sbin/
+COPY --from=debian /sbin/mke2fs /sbin/mke2fs
+COPY --from=debian /etc/mke2fs.conf /etc/mke2fs.conf 
+COPY --from=debian /sbin/mkfs* /sbin/
+COPY --from=debian /sbin/fsck* /sbin/
+COPY --from=debian /sbin/e2fsck /sbin/e2fsck
+COPY --from=debian /sbin/xfs_repair /sbin/xfs_repair
+COPY --from=debian /usr/sbin/xfs* /usr/sbin/
+COPY --from=debian /usr/include/xfs /usr/include/xfs
+COPY --from=debian /usr/lib/xfsprogs/xfs* /usr/lib/xfsprogs/
+COPY --from=debian /usr/include/xfs /usr/include/xfs
+COPY --from=debian /sbin/blkid /sbin/blkid
+# Copy shared libraries into distoless base.
+COPY --from=debian /lib/x86_64-linux-gnu/libselinux.so.1 /lib/x86_64-linux-gnu/libselinux.so.1
```

Comment:
these shared libraries are compiled for x86 from the prefix and I think that the multidistro build also builds for arm

ref: https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver/blob/bc88549c3cdc1f5881af5992d04a1e13ae314def/Makefile#L98-L102

Reply:
Interesting point. Maybe we can use the docker $BUILDPLATFORM or something like that?

I wonder if this sort of thing has been done before, maybe there's some common advice for dealing with these issues? eg, the debian base that we're pulling from presumably would be arm when such a build.

Or we just use a different dockerfile for arm, like we do for windows.

Reply:
I decided to go with a separate dockerfile for arm as opposed to checking if the environment variable TARGETPLATFORM = "arm64" because conditionals in docker (e.g. `RUN if [ "$TARGETPLATFORM" = "arm64"]; then ...`) uses bash. Since our distroless image does not have shell, we would have to use a distroless base image that has shell, but the bash image contains a much larger potential attack surface than the other distroless images. I thought about doing something like making a new folder in the debian image x86_64-linux-gnu-containerized then moving all the content in /lib/x86_64-linux-gnu/ to that folder if the TARGETPLATFORM is "arm64" since the debian image has bash. Then in the distroless image we always copy from the x86_64-linux-gnu-containerized  folder, but the docker build will fail if the COPY doesn't COPY over at least one file, which would be the case for the arm64.